File: web/src/api/admin/score.ts
import { api, callApi } from "../config";
import {
  Competition,
  Registration,
  Score,
  StudentScore,
  ApiResponse,
} from "../../types";

export interface CreateScoreRequest {
  competition_id: number;
  student_scores: StudentScore[];
}

export interface ReviewScoreRequest {
  competition_id: number;
}

/**
 * 管理员-成绩录入API
 */
export const adminScoreInputAPI = {
  /**
   * 获取比赛列表（用于成绩录入）
   */
  getCompetitions: async (): Promise<ApiResponse<Competition[]>> => {
    return await callApi(() => api.get("/admin/scores/input/competitions"));
  },

  /**
   * 创建或更新成绩
   */
  createOrUpdateScores: async (
    data: CreateScoreRequest,
  ): Promise<ApiResponse<void>> => {
    return await callApi(() => api.post("/admin/scores/input", data));
  },

  /**
   * 获取比赛成绩
   */
  getCompetitionScores: async (id: number): Promise<ApiResponse<Score[]>> => {
    return await callApi(() => api.get(`/admin/scores/input/${id}`));
  },

  /**
   * 删除成绩记录
   */
  deleteScores: async (id: number): Promise<ApiResponse<void>> => {
    return await callApi(() => api.delete(`/admin/scores/input/${id}`));
  },

  /**
   * 获取所有报名学生（用于成绩录入）
   */
  getRegisteredStudents: async (
    competitionId: number,
  ): Promise<ApiResponse<Registration[]>> => {
    return await callApi(() =>
      api.get(`/admin/scores/input/${competitionId}/registrations`),
    );
  },
};

/**
 * 管理员-成绩审核API
 */
export const adminScoreReviewAPI = {
  /**
   * 获取比赛列表（用于成绩审核）
   */
  getCompetitions: async (): Promise<ApiResponse<Competition[]>> => {
    return await callApi(() => api.get("/admin/scores/review/competitions"));
  },

  /**
   * 获取比赛成绩
   */
  getCompetitionScores: async (id: number): Promise<ApiResponse<Score[]>> => {
    return await callApi(() => api.get(`/admin/scores/review/${id}`));
  },

  /**
   * 审核成绩
   */
  reviewScores: async (
    data: ReviewScoreRequest,
  ): Promise<ApiResponse<void>> => {
    return await callApi(() => api.post("/admin/scores/review", data));
  },
};

File: api/handlers/score.go
package handlers

import (
	"net/http"
	"strconv"

	"github.com/SHXZ-OSS/sports-meeting-system/api/middlewares"
	"github.com/SHXZ-OSS/sports-meeting-system/models"
	"github.com/SHXZ-OSS/sports-meeting-system/types"
	"github.com/SHXZ-OSS/sports-meeting-system/utils"
	"github.com/gin-gonic/gin"
)

// CreateScoreRequest 创建成绩请求
type CreateScoreRequest struct {
	CompetitionID int                  `json:"competition_id" binding:"required"`
	StudentScores []types.StudentScore `json:"student_scores" binding:"required,dive"`
}

// ReviewScoreRequest 审核成绩请求
type ReviewScoreRequest struct {
	CompetitionID int `json:"competition_id" binding:"required"`
}

// GetPublicCompetitionScores 获取比赛的所有成绩（游客）
func GetPublicCompetitionScores(c *gin.Context) {
	// 解析路径参数
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		utils.ResponseError(c, http.StatusBadRequest, "无效的比赛ID")
		return
	}

	// 检查比赛是否存在
	competition, err := models.GetCompetitionByID(id)
	if err != nil {
		utils.ResponseError(c, http.StatusNotFound, "比赛不存在")
		return
	}
	// 游客只能查看status为completed的比赛成绩
	if competition.Status != types.StatusCompleted {
		utils.ResponseError(c, http.StatusForbidden, "比赛结果未出或在审核中")
		return
	}

	// 获取成绩列表
	scores, err := models.GetScoresByCompetitionID(id)
	if err != nil {
		utils.ResponseError(c, http.StatusInternalServerError, "获取成绩列表失败")
		return
	}

	// 如果指定了分页，返回分页响应
	utils.ResponseOK(c, scores)
}

// GetCompetitionScores 获取比赛的所有成绩（管理员）
func GetCompetitionScores(c *gin.Context) {
	// 解析路径参数
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		utils.ResponseError(c, http.StatusBadRequest, "无效的比赛ID")
		return
	}

	// 获取成绩信息
	score, err := models.GetScoresByCompetitionID(id)
	if err != nil {
		utils.ResponseError(c, http.StatusNotFound, "成绩记录不存在")
		return
	}

	// 返回响应
	utils.ResponseOK(c, score)
}

// GetStudentScores 获取学生成绩（学生端使用）
func GetStudentScores(c *gin.Context) {
	// 从上下文获取学生ID
	studentID, ok := middlewares.GetUserIDFromContext(c)
	if !ok {
		utils.ResponseError(c, http.StatusUnauthorized, "未授权")
		return
	}

	// 获取学生成绩
	scores, err := models.GetScoresByStudentID(studentID)
	if err != nil {
		utils.ResponseError(c, http.StatusInternalServerError, "获取成绩失败")
		return
	}

	// 返回响应
	utils.ResponseOK(c, scores)
}

// GetStudentScoresById 获取指定学生的成绩
func GetStudentScoresById(c *gin.Context) {
	// 解析路径参数
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		utils.ResponseError(c, http.StatusBadRequest, "无效的学生ID")
		return
	}

	// 获取学生成绩
	scores, err := models.GetScoresByStudentID(id)
	if err != nil {
		utils.ResponseError(c, http.StatusInternalServerError, "获取成绩失败")
		return
	}
	// 返回响应
	utils.ResponseOK(c, scores)
}

// CreateOrUpdateScores 创建或更新成绩记录
func CreateOrUpdateScores(c *gin.Context) {
	// 解析请求
	var req CreateScoreRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.ResponseError(c, http.StatusBadRequest, "无效请求")
		return
	}

	// 获取提交人ID
	submitterID, ok := middlewares.GetUserIDFromContext(c)
	if !ok {
		utils.ResponseError(c, http.StatusUnauthorized, "未授权")
		return
	}

	// 创建或更新成绩
	err := models.CreateOrUpdateScores(req.CompetitionID, req.StudentScores, submitterID)
	if err != nil {
		utils.ResponseError(c, http.StatusInternalServerError, "提交成绩失败: "+err.Error())
		return
	}

	// 返回响应
	utils.ResponseSuccessWithCustomMessage(c, "上传成功")
}

// ReviewScores 审核成绩
func ReviewScores(c *gin.Context) {
	// 解析请求
	var req ReviewScoreRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.ResponseError(c, http.StatusBadRequest, "无效请求")
		return
	}

	// 获取审核人ID
	reviewerID, ok := middlewares.GetUserIDFromContext(c)
	if !ok {
		utils.ResponseError(c, http.StatusUnauthorized, "未授权")
		return
	}

	// 审核成绩
	if err := models.ReviewCompetitionScoresByID(req.CompetitionID, reviewerID); err != nil {
		utils.ResponseError(c, http.StatusInternalServerError, "审核成绩失败: "+err.Error())
		return
	}

	// 返回响应
	utils.ResponseSuccessWithCustomMessage(c, "审核成功")
}

// DeleteScores 删除成绩记录
func DeleteScores(c *gin.Context) {
	// 解析路径参数
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		utils.ResponseError(c, http.StatusBadRequest, "无效的成绩ID")
		return
	}

	// 删除成绩记录
	if err := models.DeleteCompetitionScoresByID(id); err != nil {
		utils.ResponseError(c, http.StatusInternalServerError, "删除成绩记录失败: "+err.Error())
		return
	}

	// 返回响应
	utils.ResponseSuccessWithCustomMessage(c, "删除成功")
}
